src/diagnostics.rs

```
@@ -0,0 +1,204 @@
+        }
+    }
+
+    pub fn bug<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        let mut diag =
+            Some(self.diagnostic(Severity::Bug, message.into()).note("please report this"));
+        backtrace::trace(|frame| {
+            let ip = frame.ip();
+            backtrace::resolve(ip, |symbol| {
+                diag = Some(diag.take().unwrap().note(format!(
+                    "in heradoc file {:?} name {:?} line {:?} address {:?}",
+                    symbol.filename(),
+                    symbol.name(),
+                    symbol.lineno(),
+                    symbol.addr()
+                )));
+            });
+            true
+        });
+        diag.unwrap()
+    }
+
+    pub fn error<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Error, message.into())
+    }
+
+    pub fn warning<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Warning, message.into())
+    }
+
+    pub fn note<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Note, message.into())
+    }
+
+    pub fn help<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Help, message.into())
+    }
+}
+
+#[must_use = "call `emit` to emit the diagnostic"]
+pub struct DiagnosticBuilder<'a: 'b, 'b> {
+    file_map: &'b FileMap<&'a str>,
+    out: &'b mut StandardStream,
+    diagnostics: Vec<Diagnostic>,
+
+    severity: Severity,
+    message: String,
+    code: Option<String>,
+    labels: Vec<Label>,
+}
+
+impl<'a: 'b, 'b> DiagnosticBuilder<'a, 'b> {
+    pub fn emit(self) {
+        let Self { file_map, out, mut diagnostics, severity, message, code, labels } = self;
+        diagnostics.push(Diagnostic { severity, message, code, labels });
+
+        // ignore output errors, because where would we log them anyway?!
```

Yes, where would we log them. But wouldn't that mean that it's more sensible to `panic!()` instead of, say, continue trying to write to a full disk or buffer.

---

If we panic, that'll still try to write a message to the output, which is already gone. So the only sane way to handle this would be an `abort()`, but then the user might wonder why the program is just gone without any error message.

So there isn't really any sane way to handle this. This is something that just shouldn't happen :) I guess I should just panic here and let the `panic_handler` take care of the rest. Anyway, even this might be bad, because the panic handler writes stuff again.